### PR TITLE
Geometry: add a Geometry reader for various supported mesh files

### PR DESCRIPTION
**Pull request by tamiko**
Extracted from #130

### CODE REVIEW

File: source/geometry_gmsh.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+     */
+    template <int dim>
+    class GMSH : public Geometry<dim>
+    {
+    public:
+      GMSH(const std::string subsection)
+          : Geometry<dim>("gmsh", subsection)
+      {
+        filename_ = "ryujin.msh";
+        this->add_parameter("filename", filename_, "GMSH *.msh file to load");
```
Comment (tamiko) [inline source/geometry_gmsh.h]: @gregorychristian I have changed your commit slightly by adding a parameter for the file that should be read in and renamed the class to `GMSH`.

**Comment by ejtovar** [comment]
@tamiko @gregorychristian Is it possible to generalize this GMSH reader to include UCD meshes as well? That is, depending on the type of mesh that gets inputted, the reader decides if it is GMSH vs UCD? See: https://www.dealii.org/current/doxygen/deal.II/classGridIn.html#abbb16aa0aa53bfa9231b69278d397616

File: source/geometry_gmsh.h [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+      }
+
+      void create_triangulation(
+          typename Geometry<dim>::Triangulation &triangulation) final
+      {
+        dealii::GridIn<dim> gridin;
+        gridin.attach_triangulation(triangulation);
+        std::ifstream file(filename_);
+        gridin.read_msh(file);
```
Comment (tamiko) [inline source/geometry_gmsh.h]: The logic is so small that we can simply do the call to `GridIn` in the `create_triangulation()` function.

(The split that you copied where the implementation is split out into a `GridGenerator` function served a special purpose to be closer to the deal.II `GridGenerator` functions. But that isn't needed here.)

File: README.md [hunk 3]
```diff
@@ -132,4 +132,5 @@ Contributors
  - Seth Gerberding ([@Gerbeset](https://github.com/Gerbeset)), Texas A&M University, TX, USA
  - David Pecoraro ([@ChrisPec27](https://github.com/ChrisPec27)), Texas A&M University, TX, USA
  - Madison Sheridan ([@Helblindi](https://github.com/Helblindi)), Texas A&M University, TX, USA
+ - Gregory Christian ([@gregorychristian](https://github.com/gregorychristian), Imperial College London, UK
```
Comment (tamiko) [inline README.md:135]: @gregorychristian If you don't mind we'd like to add you as contributor to the list.

**Review by tamiko — COMMENTED** [review]
@gregorychristian I have split out the GMSH part into a separate pull request.

**Comment by tamiko** [comment]
@ejtovar Excellent point. I went ahead and refactored the class to be called `Reader` and switched to `GreadIn::read()` that supports various different mesh formats.

File: README.md (hunk 3 above)
Comment (gregorychristian) [reply README.md:135]: @tamiko Yes of course!

File: source/geometry_gmsh.h (hunk 1 above)
Comment (gregorychristian) [reply source/geometry_gmsh.h]: @tamiko Great thanks.

**Review by gregorychristian — APPROVED** [review]

**Review by gregorychristian — APPROVED** [review]

**Review by gregorychristian — APPROVED** [review]

**Pull request merged**